ci/kind/kind-setup.sh

```diff
@@ -528,19 +528,27 @@ function destroy_external_servers {
 
 function clean_kind {
     echo "=== Cleaning up stale kind clusters ==="
-    read -a all_kind_clusters <<< $(kind get clusters)
-    for kind_cluster_name in "${all_kind_clusters[@]}"; do
-        creationTimestamp=$(kubectl get nodes --context kind-$kind_cluster_name -o json -l node-role.kubernetes.io/control-plane | \
-        jq -r '.items[0].metadata.creationTimestamp')
+    LOCK_FILE="$THIS_DIR/.kube/config.lock"
+    (
+      flock -x 200
```

Comment:
@jainpulkit22 I'm a bit confused. I first I thought this would work because `.kube/config.lock` was also used by `kind` / `kubectl`. See https://github.com/kubernetes-sigs/kind/blob/1c5a56b30145d98a61c0538975ca0894ad65e0d0/pkg/cluster/internal/kubeconfig/internal/kubeconfig/lock.go#L27

But then I ran some tests and it seems to me that `flock` and the Go code used to acquire the lock (`os.OpenFile(lockName(filename), os.O_CREATE|os.O_EXCL, 0)`) are not "compatible". For example, a bash script using flock to acquire an exclusive lock using `~/.kube/config.lock` can execute concurrently which a `kind` command that acquires an excluisve lock on the same file using the code I linked to above.

This is the code I used for testing:
<details>
<summary>test.go</summary>

```go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"time"
)

// these are from
// https://github.com/kubernetes/client-go/blob/611184f7c43ae2d520727f01d49620c7ed33412d/tools/clientcmd/loader.go#L439-L440

func lockFile(filename string) error {
	// Make sure the dir exists before we try to create a lock file.
	dir := filepath.Dir(filename)
	if _, err := os.Stat(dir); os.IsNotExist(err) {
		if err = os.MkdirAll(dir, 0755); err != nil {
			return err
		}
	}
	f, err := os.OpenFile(lockName(filename), os.O_CREATE|os.O_EXCL, 0)
	if err != nil {
		return err
	}
	f.Close()
	return nil
}

func unlockFile(filename string) error {
	return os.Remove(lockName(filename))
}

func lockName(filename string) string {
	return filename + ".lock"
}

func main() {
	if err := lockFile("/tmp/.clusters"); err != nil {
		panic("Failed to lock")
	}
	fmt.Println("Sleeping...")
	time.Sleep(10 * time.Second)
	fmt.Println("Done")
	if err := unlockFile("/tmp/.clusters"); err != nil {
		panic("Failed to unlock")
	}
}
```

</details>

<details>
<summary>test.sh</summary>

```bash
(
    flock -x 200
    echo "SLEEPING..."
    sleep 10
    echo "DONE"
) 200>>/tmp/.clusters.lock
```

</details>

Could you confirm that you tested this approach successfully? Maybe there is an issue in my test programs.

If the 2 locking schemes are not compatible, I don't think this patch will have the intended effect, as we can have a concurrent call to `kind create cluster` which will mutate the Kubeconfig file?

It also seems that because `flock` will create the file and not delete it, `kind` will fail to acquire the lock later (file already exists). I tried and got this error from `kind`:
```
ERROR: failed to create cluster: failed to lock config file: open /Users/abas/.kube/config.lock: file exists
```

Finally, even if the locking schemes were compatible, I am not sure that the current version would be sufficient, as `kind create cluster` will only acquire the lock while it's writing the config: therefore, maybe we could still have a context in the config for which the control plane is not ready yet?

Reply:
I have explicitly added the command to remove the lock file, other than that it seems compatible to me, because when I ried locally b oth the scripts together the create script waited for the lock to be released by the delete script.

Regarding your second question that we could still have a context in the config for which the control plaane is not ready yet, taht won't be a problem for use because the output of creation timeSTamp is piped to true, so the script won't panic and will run as usual.

Reply:
> I have explicitly added the command to remove the lock file, other than that it seems compatible to me, because when I ried locally b oth the scripts together the create script waited for the lock to be released by the delete script.

Please share your steps, this is not what I observe in my case. `test.sh` and `test.go` can execute concurrently (`test.sh` is supposed to represent the clean up step in `kind-setup.sh` here, using the same `flock` command).

Reply:
I didn't use test.go for my local testing I did the following:
I created a test.sh script similar to the above one and executed `kind create cluster` command in one shell and `test.sh` script in another cell, and there I din't got any error.
I did this because in actual test scenario also we are just making calls to `kind create cluster` command.